web3/utils/validation.py

```diff
@@ -33,11 +35,19 @@ def validate_abi(abi):
     """
     Helper function for validating an ABI
     """
+    seen_selectors = set()
     if not is_list_like(abi):
         raise ValueError("'abi' is not a list")
     for e in abi:
         if not is_dict(e):
             raise ValueError("The elements of 'abi' are not all dictionaries")
+        if e['type'] == 'function':
+            selector = encode_hex(function_abi_to_4byte_selector(e))
+            if selector in seen_selectors:
+                raise ValueError(
+                    'Found collisions of functions with selector: %s' % selector
```

Comment:
*Ideally* this would spit out the names of all signatures that have a matching selector. There's probably a clean way to build that without iteration and mutation (maybe with the help of toolz).

Reply:
I have spit out signature names for now using a dict. However, I don't better ideas to make this concise. If you have a better idea let me know. Or even a link to some documentation of toolz would do, I'm willing to explore.

Reply:
To clarify: this only displays the name of the first two conflicting functions, rather than all of the colliding functions (say, if there were three, or two pairs of colliding function signatures). So for example: [`groupby`](http://toolz.readthedocs.io/en/latest/api.html#toolz.itertoolz.groupby) could group by selector.

Reply:
That was done intentionally. It's just my personal preference to tackle one error a time. But sure, we can spit out all possible errors together!